samcli/commands/local/lib/intrinsic_resolver/intrinsic_property_resolver.py

```diff
@@ -0,0 +1,759 @@
+        }
+
+        self.conditional_key_function_map = {
+            IntrinsicResolver.FN_AND: self.handle_fn_and,
+            IntrinsicResolver.FN_OR: self.handle_fn_or,
+            IntrinsicResolver.FN_IF: self.handle_fn_if,
+            IntrinsicResolver.FN_EQUALS: self.handle_fn_equals,
+            IntrinsicResolver.FN_NOT: self.handle_fn_not
+        }
+
+    def intrinsic_property_resolver(self, intrinsic, parent_function="template"):
+        """
+        This resolves the intrinsic of the format
+        {
+            intrinsic: dict
+        } by calling the function with the relevant intrinsic function resolver.
+
+        This also supports returning a string, list, boolean, int since they may be intermediate steps in the recursion
+        process. No transformations are done on these.
+
+        By default this will just return the item if non of the types match. This is because of the function
+        resolve_all_attributes which will recreate the resources by processing every aspect of resource.
+
+        Parameters
+        ----------
+        intrinsic: dict, str, list, bool, int
+            This is an intrinsic property or an intermediate step
+        parent_function: str
+            In case there is a missing property, this is used to figure out where the property resolved is missing.
+
+        Return
+        ---------
+        The simplified version of the intrinsic function. This could be a list,str,dict depending on the format required
+        """
+        if intrinsic is None:
+            raise InvalidIntrinsicException("Missing Intrinsic property in {}".format(parent_function))
+
+        if any(isinstance(intrinsic, object_type) for object_type in [string_types, list, bool, int]):
+            return intrinsic
+
+        if intrinsic == {}:
+            return intrinsic
+
+        if not isinstance(intrinsic, dict):
+            raise InvalidIntrinsicException(
+                "Invalid Intrinsic type. It is not a int, list, str, or dict {}".format(intrinsic))
+
+        keys = list(intrinsic.keys())
+        key = keys[0]
+
+        if key in self.intrinsic_key_function_map:
+            intrinsic_value = intrinsic.get(key)
+            return self.intrinsic_key_function_map.get(key)(intrinsic_value)
+        elif key in self.conditional_key_function_map:
+            intrinsic_value = intrinsic.get(key)
+            return self.conditional_key_function_map.get(key)(intrinsic_value)
+
+        # In this case, it is a dictionary that doesn't directly contain an intrinsic resolver and must be
+        # re-parsed to resolve.
+        sanitized_dict = {}
+        for key, val in intrinsic.items():
+            sanitized_key = self.intrinsic_property_resolver(key, parent_function=parent_function)
+            sanitized_val = self.intrinsic_property_resolver(val, parent_function=parent_function)
+            verify_intrinsic_type_str(sanitized_key,
```

Comment:
Fail after each invalid intrinsic might not be a good user experience. Imagine you are a customers that has 5 different 'bugs' with intrinsics in your template. In order to make a single invoke work, you will be running this 5 times. Could be worth adding a collector that captures errors and prints them all at once.

Reply:
That is a good idea when dealing with errors. One issue with this is with objects that are looking for a certain type. For example, Fn::Select requires the first argument to be an integer. If there is an error during nested resolution, can we continue execution to resolve further errors? This could also possibly be issues with templates that point to deployed cloud resources.